ruby_on_rails/testing/model_testing.md

```diff
@@ -0,0 +1,125 @@
+
+~~~
+
+class EmailValidator < ActiveModel::EachValidator
+  def validate_each(record, attribute, value)
+    unless value =~ /\A([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})\z/i
+      record.errors.add attribute, (options[:message] || "is not an email")
+    end
+  end
+end
+
+class User < ApplicationRecord
+  validates :email, presence: true, email: true
+end
+
+~~~
+ 
+In the above code we add our own validator to the email validation. [Validate each](https://guides.rubyonrails.org/active_record_validations.html#custom-validators). By using validate the model will call on the validator to ensure the presence of valid characters. That then brings us to the following RSpec test where we will write the context and what we expect the outcome to be. The validate_each method is also where you would make changes as well if the test failed to pass.
+
+~~~
+
+RSpec.describe User, type: :model do
+  describe '#valid?' do
+    context "when the user does not have an email address" do
+      user =  build(:user)
+
+      it "will not be valid" do
+        expect(user.valid?).to be false
+      end
+    end
+    
+    context "when the user has an email address" do
+      user =  build(:user, email_address: "jdoe@example.com")
+
+      it "will be valid" do
+        expect(user.valid?).to be true 
+      end
+    end
+  end
+end
+
+~~~ 
+
+One thing to note is that we have two tests. We want to test both that it only accepts valid characters while having a separate test for rejecting the invalid ones. The other thing to note is that you can set attributes when creating a factory record as well rather than going back to the factory file. The Factory bot [repository](https://github.com/thoughtbot/factory_bot/blob/main/GETTING_STARTED.md#setup) has excellent information for diving in deeper and it's highly reccomended to read through it to understand factory bot better. 
+
+### Knowledge Check
+
+This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.
+
+*   <a class="knowledge-check-link" href="#what-are-model-tests">What are model tests?</a>
+*   <a class="knowledge-check-link" href="#creating-a-factory">How would I create a factory to use in my tests?</a>
+*   <a class="knowledge-check-link" href="#example-model-test">What are some examples of testing business logic?</a>
+
+### Additional Resources
+
+This section contains helpful links to related content. It isn’t required, so consider it supplemental.
+
+*   For an example of a model test and having a useful resource. You can visit [code with Jason](https://www.codewithjason.com/write-model-tests-part/).
+*   For an alternative if RSpecs built in database cleaner isn't working for testing. You can find it [here](https://github.com/DatabaseCleaner/database_cleaner).
```

Comment:
I prefer an article that explains more about the need for database cleaner, then just providing a link to the gem. I liked this article, but I am sure there are others: https://thoughtbot.com/blog/how-we-test-rails-applications#database-cleaner. 

If you disagree, let me know why you prefer this approach.

Reply:
I really like this article. It has some solid links in the section plus it was updated last year so fairly relevant vs a 10 year old article.